data/blog/how-to-refactor-and-not-break-things.md

```diff
@@ -0,0 +1,101 @@
+
+This way we had a couple of smaller PRs that each dealt with one topic and were easy to review.
+
+After we moved all the functionality from the Hub into the Scope our extensive test suite was still all green, giving us confidence that we did not change any behavior. Because neither the top-level Sentry API nor the Hub-based API was changed this could be released in a minor version update.
+
+This concluded the preparation of our canvas. Time to make the change.
+
+## Phase II: New Scopes
+
+Phase I left us with a hollow shell of a Hub and all functionality in the Scope. We now refactored the Scope. This was the biggest part. We changed how the Scope was stored in memory (it's not on the Hub anymore but saved as a Python Context Variable). We also introduced three different flavors of the Scope for better encapsulation of data. If you want to dig into details read our [develop docs on the topic](https://develop.sentry.dev/sdk/hub_and_scope_refactoring/).
+
+After we updated the Scope the old Hub-based API could still be used, but under the hood, the new Scopes-based API was called. We did not release this phase because we wanted to first do Phase III.
+
+Again, our test suite gave us confidence that the SDK still behaved the same as before.
+
+## Phase III: Use New Scopes Everywhere
+
+In the Python SDK, we have over 40 integrations for various web frameworks, databases, and other libraries. All of them still used the Hub-based API.
+
+We updated each integration from the old Hub-based API to the new Scope-based API in a separate PR. This was pretty straightforward and done in a couple of days.
+
+Doing this gave us insights into the look and feel of the new API. Because we now used the API in the same way as our power users would use it in the future. We found some things that we did not like and made some minor changes to the new API to make it more convenient to use. Having one PR for each integration created a lot of small PRs that were easy to review.
+
+After we had all PRs merged, we created a release candidate and released it on PyPI so everyone could give it a try and we could gather feedback.
+
+We updated our internal usage dashboards so we could see the adoption of the new release candidate.
+
+## Phase IV: Load Testing and Dogfooding
+
+We started to do load testing. We created a [sample application and ran load tests](https://github.com/getsentry/demo-flask-load-test) locally to check if the CPU and memory usage would change between the current SDK and the new 2.0 version. It did not. After the load tests, we were confident that we could use the release candidate on Sentry.io.
+
+Sentry.io is a very big Python project where we use all the advanced features of the SDK. It is a perfect candidate for dogfooding because it will uncover problems very soon.
```

Comment:
```suggestion
[Sentry.io is a very big Python code base](https://github.com/getsentry/sentry) where we use all the advanced features of the SDK. It is a perfect candidate for dogfooding because it will uncover problems very soon.
```